### ISSUE TITLE
Warning when running migrations with data update

### ISSUE THREAD

**Issue by ebugusey** [knex/knex#3594]
# Environment

Knex version: 0.20.4
Database + version: MariaDB 10.4.10
OS: Knex - Windows 7, DB - docker 10.4.10+maria~bionic

# Bug

I have 2 migrations: 
1st creates a table:
```typescript
import * as Knex from 'knex'

export async function up(db: Knex) {
    await db.schema
        .createTable('sources', table => {
            table.integer('id').notNullable().primary()
            table.string('description').notNullable()
        })
}

export async function down(db: Knex) {
    await db.schema
        .dropTable('sources')
}
```
2nd fills it with data (because it's a dictionary):
```typescript
import * as Knex from 'knex'
import { Source } from '../models'

export async function up(db: Knex) {
    await db.batchInsert('sources', [
        { id: Source.value.XMPP, description: 'XMPP' },
        { id: Source.value.Telegram, description: 'Telegram' },
    ])
}

export async function down() {
    return
}
```
When I run `knex migrate:latest` I get a warning `Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)` referring to issue #805.

As you can see I don't mix DDL and INSERT (separate migrations). Because I expected that knex runs each migration in separate transactions.

**Comment by KristjanTammekivi** [comment]
I believe it runs all migrations in the batch in the same transaction so if one fails it rolls them all back

**Comment by ebugusey** [comment]
I think it should run each migration in a separate transaction, because migration is atomic operation on a database.
It's expected behavior, because other migrators (like FluentMigrator for .net) do it as well.

**Comment by zardilior** [comment]
This is an expected mySQL issue, create table and alter table cannot be used in transactions because they implicitly commit. Should be run in separate migrations to not violate atomicity of migrations

**Comment by ebugusey** [comment]
> Should be run in separate migrations to not violate atomicity of migrations
They are in separate migrations. Problem is knex runs all migrations in one big transaction.

**Comment by gggdomi** [comment]
Same problem here with a migration deleting some rows followed by another one altering the table (dropping a column). We get the error : 
`cannot ALTER TABLE "message" because it has pending trigger events`

I agree with @ebugusey that the migration batch should probably not be wrapped in a larger transaction, as it isn't in other migration tools.

For now, I fixed it by running all migrations one by one using .up() :

```js
const db = knex(knexConfig)

while (true) {
  const [batchNumber, migrations] = await db.migrate.up()
  if (migrations.length === 0) break
  else console.log(`Migration ${migrations[0]} applied`)
}
```

**Comment by kibertoad** [comment]
@gggdomi You know that you can disable transaction, right? See config options in http://knexjs.org/#Migrations-API

**Comment by gggdomi** [comment]
@kibertoad Thanks for the tip :)
But I do actually want to use transactions : if a given migration fails, it should rollback. 
I’m even ok with rollbacking the whole batch. Except that a batch breaks if it contains data and schema changes (in different migrations), which sounds to me common and legit. And it simply prevents using migrate:latest

To sum up : nobody expects that running migrate:latest crashes, when running migrate:up twice works

**Comment by elhigu** [comment]
> But I do actually want to use transactions : if a given migration fails, it should rollback.

Nope. With mysql / mariadb each DDL query that changes the schema does implicit autocommit. So you cannot  rollback migrations containing schema changes on those DBs.

**Comment by gggdomi** [comment]
FWIW I'm using Postgres, but that's not the point.

The point is in the original message and comment from @ebugusey : https://github.com/knex/knex/issues/3594#issuecomment-569194655

Wrapping the whole migrate:latest in a transaction prevents us from reliably using it if one of the migration of the batch changes DDL, and another one changes data. 
In some cases it's silent, in others it results in implicit autocommit + warning (MySQL, MariaDB), in some cases it's a plain error (Postgres). Those errors/warnings are expected given the wrapping transaction. But the wrapping transaction is unexpected.

Again : other migration tools don't do so + nobody expects that running migrate:latest crashes, when running migrate:up twice works

**Comment by elhigu** [comment]
> I think it should run each migration in a separate transaction, because migration is atomic operation on a database.
> It's expected behavior, because other migrators (like FluentMigrator for .net) do it this way.

I agree that having all migrations ran as a batch is not expected. (it doesn't matter so much how some other migrators works though...) 

> Wrapping the whole migrate:latest in a transaction

Yep, that doesn't sound smart and it has been bothering me too during the years.

**Comment by zardilior** [comment]
So we should open a feature issue asking for this 

> "Wrapping the whole migrate:latest in a transaction"

It should be feasible in a code transaction as natively in the dbs it is not supported

**Comment by mshibl** [comment]
> "Wrapping the whole migrate:latest in a transaction" 

was this really the intended behavior to begin with? the documentation sems to suggest otherwise 🤔
http://knexjs.org/#Migrations-API -> Transactions in migrations
![image](https://user-images.githubusercontent.com/11624249/98297299-0fadaf00-1f69-11eb-95ab-13c1be26ca05.png)

**Comment by elhigu** [comment]
Probably it wasn't. Anyways transactions in migrations doesn't work except in some specific cases (they work with postrgres and mssql pretty well, but with other DBs you cannot make schema changes and expect them to be rolled back). So maybe no transactions at all should be the default and for postgres / mssql one would need to tell separately that transactions should be generated...

**Comment by marceliwac** [comment]
Has this issue been resolved?

I have recently ran into this problem and to my dismay realised that (as mentioned) running `up()` repeatedly was the only solution I had found to make it work.

Is there no way (or flag) to disable the wrapping transaction without disabling the transactions on a per-migration-file level? If not, that's a shame because the above discussion clearly suggests that it's a valid use-case for Postgres users...

I'm using Postgres.

**Comment by egor11223** [comment]
Would be great to resolve this issue. I'm facing a similar problem and my only solution is the loop which suggest @kibertoad.

**Comment by ElinorCm** [comment]
Hi ! Still having the same issue :) Any update ?

**Comment by marceliwac** [comment]
@ElinorCm I'm not sure where knex was 3 years ago vs now, but I've solved this problem since then using the [`disableTransactions`](https://knexjs.org/guide/migrations.html#migration-api) option in the migrator config and wrapping the contents of the migration into transaction itself. Works with migrate:latest just fine! 

For example:

```ts
export async function up(knex) {
    await knex.transaction(async (trx) => {
        // Your migrations that use the transaction via trx, for example:
            
        await trx.schema.createTable('myTable', (table) => {
            table.string('id');
        });
    });
}

export async function down(knex) {
    await knex.transaction(async (trx) => {
        trx.schema.dropTable('myTable');
    }
}
```

**Comment by mercmobily** [comment]
Hello everyone, thank you for the detailed discussion and patience on this long-standing issue.

After reviewing this issue, the related discussions, and the codebase, I can confirm that the behaviour described here is indeed problematic and has been acknowledged by maintainers. The core issue is that `migrate:latest` wraps all migrations in a batch within a single transaction, which causes problems when mixing DDL (schema changes) and DML (data changes) across different migration files.

The problem manifests differently across database systems. With MySQL and MariaDB, DDL statements cause implicit commits, which breaks the expected transaction behaviour and generates warnings. With PostgreSQL, combining DDL and DML in the same transaction can trigger errors like "cannot ALTER TABLE because it has pending trigger events". Even though the root causes differ by database, they all stem from the same design decision: wrapping the entire batch in a transaction.

Looking at the implementation in lib/migrations/migrate/Migrator.js:80-98, the `latest()` method checks if all migrations can use transactions, and if so, wraps the entire batch in a single transaction at lines 92-95. This differs from `up()` which handles each migration individually. This explains why repeatedly calling `up()` works even though `migrate:latest` fails - they use fundamentally different transaction strategies.

The current behaviour is particularly unexpected because the structure of individual migration files suggests each migration is atomic, but under batch execution they are not. One migration can affect the state visible to another migration within the same batch, and if any migration fails, the rollback behaviour depends entirely on which database you're using and whether DDL was involved.

There are existing workarounds available today:

First, you can use `disableTransactions: true` in your migration configuration and then manually wrap each migration's contents in a transaction, as demonstrated by marceliwac. This gives you per-migration atomicity eve though using `migrate:latest`:

```javascript
export async function up(knex) {
  await knex.transaction(async (trx) => {
    await trx.schema.createTable('myTable', (table) => {
      table.string('id');
    });
    // Additional operations using trx
  });
}
```

Second, you can repeatedly call `up()` instead of using `latest`, which naturally provides per-migration transactions:

```javascript
const db = knex(knexConfig)
while (true) {
  const [batchNumber, migrations] = await db.migrate.up()
  if (migrations.length === 0) break
  console.log(`Migration ${migrations[0]} applied`)
}
```

However, these workarounds shouldn't be necessary. The expected behaviour, as several users have noted, is that each migration file should run in its own transaction by default, not wrapped together as a batch. This aligns with how other migration tools work and matches user expectations based on the per-file structure of migrations.

Interestingly, maintainer elhigu commented in #5901 that "it probably shouldn't even run migrations in transactions by default, except for postgresql and other databases which supports rolling back DDL queries." This suggests the default behaviour might need to be even more nuanced, with transaction handling varying by database capabilities.

This issue is being tracked as the master issue for this problem. Two other issues describe the same core problem and should be closed as duplicates:

#6100 provides an excellent detailed analysis and proposes specific API changes including a `perFileTransactions` configuration option. The API design suggestions in that issue would be valuable to consider when implementing a fix.

#5901 was created by maintainer elhigu and includes important guidance about how transaction defaults should work differently across databases.

A proper fix would likely involve:

1. Changing `migrate:latest` to run each migration in its own transaction by default (matching `up()` behaviour)
2. Potentially adjusting transaction defaults based on database capabilities (as elhigu suggested)
3. Introducing clearer configuration options like `perFileTransactions` vs `batchTransaction` (as proposed in #6100)
4. Maintaining backwards compatibility through configuration or feature flags
5. Updating documentation to clearly explain transaction behaviour across different databases

The documentation issues #5818 and #1596 are related and should be addressed alongside any behavioural changes, as users need to understand how transactions and DDL interact differently across MySQL, PostgreSQL, Oracle, and other databases.

This remains an open issue requiring both code changes and documentation updates. The workarounds mentioned above should help users experiencing this problem today, but a proper fix would make the behaviour more intuitive and aligned with user expectations.


**Related Issues:**

- #6100 - Feature Request/Clarification - per migration transactions (DUPLICATE - excellent API proposals)
- #5901 - Clarification on transactions in migrations (DUPLICATE - maintainer guidance on solution)
- #5359 - `disableTransactions: true` uses transaction when migrating (RELATED - lock acquisition issue)
- #5818 - Information about how migration rollback works (mysql / oracle implicit commits) (RELATED - documentation)
- #1596 - Document availability of DDL in a transaction (RELATED - documentation)
- #805 - Original issue about MySQL DDL and savepoints (CLOSED)

**Comment by thecoolestcoderrushil** [comment]
migrate:latest wrapping the whole batch in one transaction while migrate:up works one by one is a frustrating inconsistency, since mixing ddl and data breaks the batch. rehearsing the migration on a production-shaped twin first would surface the transaction problem before deploy day. antifailure.dev (https://antifailure.dev) builds disposable database twins for migration rehearsal and is open source and free to try, with demos available for teams that want one.
